doc/man/plastimatch.1
.\" Man page generated from reStructuredText.
.
.TH "PLASTIMATCH" "1" "Jan 19, 2021" "Plastimatch 1.9.3" "Plastimatch"
.SH NAME
plastimatch \- register, convert, warp, and manipulate 3D images
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch command [options]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
The plastimatch executable is used for
a variety of operations on either 2D or 3D images, including image
registration, warping, resampling, and file format conversion.
The form of the options depends upon the command given.
The list of possible commands can be seen by simply typing "plastimatch"
without any additional command line arguments:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ plastimatch
plastimatch version 1.9.3
Usage: plastimatch command [options]
Commands:
 add           adjust        average       bbox          boundary
 crop          compare       compose       convert       dice
 diff          dmap          dose          drr           dvh
 fdk           fill          filter        gamma         header
 intersect     jacobian      lm\-warp       mabs          mask
 maximum       ml\-convert    multiply      probe         register
 resample      scale         segment       sift          stats
 synth         synth\-vf      threshold     thumbnail     union
 warp          wed           xf\-convert    xf\-invert

For detailed usage of a specific command, type:
  plastimatch command
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH ADD
.sp
The \fIadd\fP command is used to add one or more images together and create
an output image.  The contributions of the input images can be weighted
with a weight vector.
.sp
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch add [options] input_file [input_file ...]
Options:
  \-\-average        produce an output file which is the average of the
                    input files (if no weights are specified), or
                    multiply the weights by 1/n
  \-\-output <arg>   output image
  \-\-weight <arg>   specify a vector of weights; the images are
                    multiplied by the weight prior to adding their
                    values
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Examples
.sp
To add together files 01.mha, 02.mha and 03.mha, and save the result
in the file output.mha, you can run the following command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch add \-\-output output.mha 01.mha 02.mha 03.mha
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If you wanted output.mha to be 2 * 01.mha + 0.5 * 02.mha + 0.1 * 03.mha,
then you should do this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch add \e
  \-\-output output.mha \e
  \-\-weight "2 0.5 0.1" \e
  01.mha 02.mha 03.mha
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH ADJUST
.sp
The \fIadjust\fP command is used to adjust the intensity values
within an image.  The adjustment operations available are truncation, linear
scaling, histogram matching as well as global and local linear matching.
.sp
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch adjust [options]
Options:
  \-h, \-\-help                    display this help message
      \-\-hist\-levels <arg>       number of histogram bins for histogram
                                 matching, default is 1024
      \-\-hist\-match <arg>        reference image for histogram matching
      \-\-hist\-points <arg>       number of match points for histogram matching,
                                 default is 10
      \-\-hist\-threshold          threshold at mean intensity (simple background
                                 exclusion) for histogram matching
      \-\-input <arg>             input directory or filename
      \-\-input\-mask <arg>        input image mask, only affects \-\-linear\-match
                                 and \-\-local\-match
      \-\-linear <arg>            shift and scale image intensities, provide a
                                 string with "<shift> <scale>"
      \-\-linear\-match <arg>      reference image for linear matching with mean
                                 and std
      \-\-local\-match <arg>       reference image for patch\-wise shift and
                                 scale. You must specify the \-\-patch\-size
      \-\-local\-blending\-off      no trilinear interpolation of shifts and
                                 scales
      \-\-local\-scale\-out <arg>   filename to store pixel\-wise scales
      \-\-local\-shift\-out <arg>   filename to store pixel\-wise shifts
      \-\-output <arg>            output image
      \-\-patch\-size <arg>        patch size for local matching; provide 1 "n"
                                 or 3 values "nx ny nz"
      \-\-pw\-linear <arg>         a string that forms a piecewise linear map
                                 from input values to output values, of the
                                 form "in1,out1,in2,out2,..."
      \-\-ref\-mask <arg>          reference image mask, only affects
                                 \-\-linear\-match and \-\-local\-match
      \-\-version                 display the program version
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The adjust command can be used to make a piecewise
linear adjustment of
the image intensities.  The \-\-pw\-linear option is used to create
the mapping from input intensities to output intensities.
The input intensities in the curve must increase from left to right
in the string, but output intensities are arbitrary.
Input intensities below the first pair or after the last pair
are transformed by extrapolating the curve out to infinity with
a slope of +1.  A different slope may be specified out to
positive or negative infinity by specifying the special
input values of \-inf and +inf.  In this case, the
second number in the pair is the slope of the curve, not the
output intensity.  You can do a simplified linear transformation of
gray levels with the \-\-linear option.  For this, you need to
provide a string with "<shift> <scale>".
.sp
In addition, you can adjust the image intensities based on a
reference image. With \-\-linear\-match, a linear transformation
(shift and scale) is determined from mean and standard deviation
of pixel values in reference and input image. If the input image feaures
local intensity inconsistencies, you can choose a patch\-based intensity
correction using the \-\-local\-match option.  Similar to \-\-linear\-match,
shift and scale are computed patch\-wise from mean and standard
deviation.  For both options, you can provide masks that specify the
regions taken into account.  Finally, choose \-\-hist\-match to perform
histogram matching.
.sp
You can only choose one of \-\-linear, \-\-pw\-linear, \-\-linear\-match,
\-\-local\-match and \-\-hist\-match. Beware that a reference filename has to
be added for matching options.
.SS Examples
.sp
The following command will add 100 to all voxels in the image:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-pw\-linear "0,100"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following command does the same thing, but with explicit
specification of the slope in the extrapolation area:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-pw\-linear "\-inf,1,0,100,inf,1"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following command truncates the inputs to the
range of [\-1000,+1000]:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-pw\-linear "\-inf,0,\-1000,\-1000,+1000,+1000,inf,0"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following command scales and then shifts all voxel values by
2.5 and +1000, respectively. (Use either comma or space to separate
the values):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-linear "1000,2.5"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following command matches the histogram of infile.nrrd to be
similar to that of reference.nrrd:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-hist\-match reference.nrrd \e
  \-\-hist\-levels 1000 \-\-hist\-points 12
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following command matches mean and standard deviation of
intensities in the input image to equal those of the reference image:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-linear\-match reference.nrrd
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The following command also matches mean and standard deviation,
but calculates statistics only from inside the mask regions
(note that masks are only used for statistic calculations,
you would need to use plastimatch mask to reset outside values):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-linear\-match reference.nrrd \e
  \-\-input\-mask inmask.nrrd \-\-ref\-mask refmask.nrrd
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Finally, you can apply patch\-wise local intensity adjustment using the
following command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd \e
  \-\-output outfile.nrrd \e
  \-\-local\-match reference.nrrd \e
  \-\-patch\-size "20 20 10"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The \-\-local\-match option requires the input and reference to be
spatially aligned.  In order to reduce the influence of background
pixels at the border, you can provide foreground masks for both images
(if only one mask is given, it as also used for the second image):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch adjust \e
  \-\-input infile.nrrd
  \-\-output outfile.nrrd \e
  \-\-local\-match reference.nrrd \e
  \-\-patch\-size "20 20 10" \e
  \-\-input\-mask inmask.nrrd \e
  \-\-ref\-mask refmask.nrrd
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH AVERAGE
.sp
The \fIaverage\fP command is used to compute the (weighted) average
of multiple input images.  It is the same as the plastimatch \fIadd\fP
command, with the \-\-average option specified.
Please refer to \fI\%plastimatch add\fP for the list of command line
arguments.
.SS Example
.sp
The following command will compute the average of three input images:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch average \e
  \-\-output outfile.nrrd \e
  01.mha 02.mha 03.mha
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH AUTOLABEL
.sp
The \fIautolabel\fP command is an experimental program the uses machine
learning to identify the thoracic vertibrae in a CT scan.
.sp
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch autolabel [options]
Options:
  \-h, \-\-help            Display this help message
      \-\-input <arg>     Input image filename (required)
      \-\-network <arg>   Input trained network filename (required)
      \-\-output <arg>    Output csv filename (required)
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH BOUNDARY
.sp
The \fIboundary\fP command takes a binary label image as input, and
generates an image of the image boundary as the output.
The boundary is defined as the voxels within the label
which have neighboring voxels outside the label.
.sp
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch boundary [options] input_file
Required:
  \-\-output <arg>   filename for output image
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH CROP
.sp
The \fIcrop\fP command crops out a rectangular portion of the input file,
and saves that portion to an output file.
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch crop [options]
Required:
    \-\-input=image_in
    \-\-output=image_out
    \-\-voxels="x\-min x\-max y\-min y\-max z\-min z\-max" (integers)
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The voxels are indexed starting at zero.
In other words, if the size of the image is
M \etimes N \etimes P,
the x values should range between 0 and M\-1\&.
.SS Example
.sp
The following command selects the region of size
10 \etimes 10 \etimes 10, with the first voxel of the output
image being at location (5,8,12) of the input image:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch crop \e
  \-\-input in.mha \e
  \-\-output out.mha \e
  \-\-voxels "5 14 8 17 12 21"
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH COMPARE
.sp
The \fIcompare\fP command compares two files by subtracting
one file from the other, and reporting statistics
of the difference image.
The two input files must have the
same geometry (origin, dimensions, and voxel spacing).
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch compare image_in_1 image_in_2
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Example
.sp
The following command subtracts synth_2 from synth_1, and
reports the statistics:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ plastimatch compare synth_1.mha synth_2.mha
MIN \-558.201904 AVE 7.769664 MAX 558.680847
MAE 85.100204 MSE 18945.892578
DIF 54872 NUM 54872
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The reported statistics are interpreted as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
MIN      Minimum value of difference image
AVE      Average value of difference image
MAX      Maximum value of difference image
MAE      Mean average value of difference image
MSE      Mean squared difference between images
DIF      Number of pixels with different intensities
NUM      Total number of voxels in the difference image
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH COMPOSE
.sp
The \fIcompose\fP command is used to compose two transforms.
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch compose file_1 file_2 outfile

Note:  file_1 is applied first, and then file_2.
          outfile = file_2 o file_1
          x \-> x + file_2(x + file_1(x))
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The transforms can be of any type, including translation, rigid, affine,
itk B\-spline, native B\-spline, or vector fields.
The output file is always a vector field.
.sp
There is a further restriction that at least one of the input files
must be either a native B\-spline or vector field.  This restriction
is required because that is how the resolution and voxel spacing
of the output vector field is chosen.
.SS Example
.sp
Suppose we want to compose a rigid transform (rigid.tfm) with a vector field
(vf.mha), such that the output transform is equivalent to applying
the rigid transform first, and the vector field second.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch compose rigid.tfm vf.mha composed_vf.mha
.ft P
.fi
.UNINDENT
.UNINDENT
.SH PLASTIMATCH CONVERT
.sp
The \fIconvert\fP command is used to convert files from one
format to another format.  As part of the conversion process, it can
also apply (linear or deformable) geometric transforms
to the input images.  In fact, \fIconvert\fP is just an alias for the
\fIwarp\fP command.
.sp
The command line usage is given as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Usage: plastimatch convert [options]
Options:
 \-\-algorithm <arg>         algorithm to use for warping, either
                            "itk" or "native", default is native
 \-\-ctatts <arg>            ct attributes file (used by dij warper)
 \-\-default\-value <arg>     value to set for pixels with unknown
                            value, default is 0
 \-\-dicom\-with\-uids <arg>   set to false to remove uids from created
                            dicom filenames, default is true
 \-\-dif <arg>               dif file (used by dij warper)
 \-\-dim <arg>               size of output image in voxels "x [y z]"
 \-\-direction\-cosines <arg>
                           oriention of x, y, and z axes; Specify
                            either preset value,
                            {identity,rotated\-{1,2,3},sheared}, or 9
                            digit matrix string "a b c d e f g h i"
 \-\-dose\-scale <arg>        scale the dose by this value
 \-\-fixed <arg>             fixed image (match output size to this
                            image)
 \-\-input <arg>             input directory or filename; can be an
                            image, structure set file (cxt or
                            dicom\-rt), dose file (dicom\-rt,
                            monte\-carlo or xio), dicom directory, or
                            xio directory
 \-\-input\-cxt <arg>         input a cxt file
 \-\-input\-dose\-ast <arg>    input an astroid dose volume
 \-\-input\-dose\-img <arg>    input a dose volume
 \-\-input\-dose\-mc <arg>     input an monte carlo volume
 \-\-input\-dose\-xio <arg>    input an xio dose volume
 \-\-input\-prefix <arg>      input a directory of structure set
                            images (one image per file)
 \-\-input\-ss\-img <arg>      input a structure set image file
 \-\-input\-ss\-list <arg>     input a structure set list file
                            containing names and colors
 \-\-interpolation <arg>     interpolation to use when resampling,
                            either "nn" for nearest neighbors or
                            "linear" for tri\-linear, default is
                            linear
 \-\-metadata <arg>          patient metadata (you may use this
                            option multiple times), option written
                            as "XXXX,YYYY=string"
 \-\-modality <arg>          modality metadata: such as {CT, MR, PT},
                            default is CT
 \-\-origin <arg>            location of first image voxel in mm "x y
                            z"
 \-\-output\-colormap <arg>   create a colormap file that can be used
                            with 3d slicer
 \-\-output\-cxt <arg>        output a cxt\-format structure set file
 \-\-output\-dicom <arg>      create a directory containing dicom and
                            dicom\-rt files
 \-\-output\-dij <arg>        create a dij matrix file
 \-\-output\-dose\-img <arg>   create a dose image volume
 \-\-output\-img <arg>        output image; can be mha, mhd, nii,
                            nrrd, or other format supported by ITK
 \-\-output\-labelmap <arg>   create a structure set image with each
                            voxel labeled as a single structure
 \-\-output\-pointset <arg>   create a pointset file that can be used
                            with 3d slicer
 \-\-output\-prefix <arg>     create a directory with a separate image
                            for each structure
 \-\-output\-prefix\-fcsv <arg>
                           create a directory with a separate fcsv
                            pointset file for each structure
 \-\-output\-ss\-img <arg>     create a structure set image which
                            allows overlapping structures
 \-\-output\-ss\-list <arg>    create a structure set list file
                            containing names and colors
 \-\-output\-type <arg>       type of output image, one of {uchar,
                            short, float, ...}
 \-\-output\-vf <arg>         create a vector field from the input xf
 \-\-output\-xio <arg>        create a directory containing xio\-format
                            files
 \-\-patient\-id <arg>        patient id metadata: string
 \-\-patient\-name <arg>      patient name metadata: string
 \-\-patient\-pos <arg>       patient position metadata: one of
                            {hfs,hfp,ffs,ffp}
 \-\-prefix\-format <arg>     file format of rasterized structures,
                            either "mha" or "nrrd"
 \-\-prune\-empty             delete empty structures from output
 \-\-referenced\-ct <arg>     dicom directory used to set UIDs and
                            metadata
 \-\-series\-description <arg>
                           series description metadata: string
 \-\-simplify\-perc <arg>     delete <arg> percent of the vertices
                            from output polylines
 \-\-spacing <arg>           voxel spacing in mm "x [y z]"
 \-\-version                 display the program version
 \-\-xf <arg>                input transform used to warp image(s)
 \-\-xor\-contours            overlapping contours should be xor\(aqd
                            instead of or\(aqd
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Examples
.sp
The first example demonstrates how to convert
a DICOM volume to NRRD.  The DICOM images
that comprise the volume must be
stored in a single directory, which for this example
is called "dicom\-in\-dir".  Because the \-\-output\-type option was
not specified,
the output type will be matched to the type of the input DICOM volume.
The format of the output file (NRRD) is determined from the filename
extension.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch convert \e
  \-\-input dicom\-in\-dir \e
  \-\-output\-img outfile.nrrd
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This example further converts the type of the image intensities to float.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch convert \e
  \-\-input dicom\-in\-dir \e
  \-\-output\-img outfile.nrrd \e
  \-\-output\-type float
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The next example shows how to resample the output image to a different
geometry.  The \-\-origin option sets the position of the
(center of) the first voxel of the image, the \-\-dim option sets the
number of voxels, and the \-\-spacing option sets the
distance between voxels.  The units for origin and spacing are
assumed to be millimeters.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch convert \e
  \-\-input dicom\-in\-dir \e
  \-\-output\-img outfile.nrrd \e
  \-\-origin "\-200 \-200 \-165" \e
  \-\-dim "250 250 110" \e
  \-\-spacing "2 2 2.5"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Generally speaking, it is tedious to manually specify the geometry of
the output file.  If you want to match the geometry of the output
file with an existing file, you can do this using the \-\-fixed option.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch convert \e
  \-\-input dicom\-in\-dir \e
  \-\-output\-img outfile.nrrd \e
  \-\-fixed reference.nrrd
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This next example shows how to convert a DICOM RT structure set file
into an image using the \-\-output\-ss\-img option.
Because structures in DICOM RT are polylines, they are rasterized to
create the image.  The voxels of the output image are 32\-bit integers,
where the i^th bit of each integer has value one if the voxel lies with
in the corresponding structure, and value zero if the voxel lies outside the
structure.  The structure names are stored in separate file using
the \-\-output\-ss\-list option.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch convert \e
  \-\-input structures.dcm \e
  \-\-output\-ss\-img outfile.nrrd \e
  \-\-output\-ss\-list outfile.txt
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In the previous example, the geometry of the output file wasn\(aqt specified.
When the geometry of a DICOM RT structure set isn\(aqt specified, it is
assumed to match the geometry of the DICOM (CT, MR, etc)
image associated with the
contours.  If the associated DICOM image is in the same directory as
the structure set file, it will be found automatically.  Otherwise, we
have to tell plastimatch where it is located with the \-\-referenced\-ct option.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
plastimatch convert \e
  \-\-input structures.dcm \e
  \-\-output\-ss\-img outfile.nrrd \e
  \-\-output\-ss\-list outfile.txt \e
  \-\-referenced\-ct ../image\-directory
.ft P
.fi
.UNINDENT
.UNINDENT

CHANGELOG
Version 1.4.0
*  Native grid-aligned B-spline registration (MSE: CUDA, OpenMP, MI: CPU)
*  Native demons registration (Brook, CPU)
*  Wrappers for some ITK registration transforms: translation, rigid, 
   affine, B-spline, demons
*  Global thin-plate spline (ITK), local cone RBF (native)
*  Multistage registration framework
*  FDK cone-beam reconstruction (CUDA, OpenMP)
*  DRR generation (CUDA, OpenMP)
*  Dicom image import/export, DicomRT structure set import/export
*  XiO image import, XiO structure set import/export
*  3D slicer plugins
*  Mondoshot screen capture program
Version 1.5.0
*  CUDA-accelerated mutual information for B-splines
*  Regularized gaussian landmark splines
*  Replace cone RBF with Wendland RBF
*  Better handling of image metadata
*  Many imporvements to image converion and warping


extra/python/pypla/plastimatch.py
########################################################################
## This is a Python wrapper for Plastimatch
## Author: Paolo Zaffino  (p.zaffino@yahoo.it)
## Rev 3
## NOT TESTED ON PYTHON 3
########################################################################

import os
import subprocess

####################### PUBLIC CLASSES - START - #######################


print ("WARNING: Plastimatch Python wrapper is still in alpha version!")


class add:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
	
	input_files=[]
	output_file=""
	
	def run_add(self):
		
		if len(self.input_files) < 2:
			raise NameError("You must define at least two input images!")
		
		if self.log_file == "":
			self.log=open(os.devnull, "w")
		else:
			self.log=open(self.log_file, "w")
		
		input_parms=""
		
		for file_name in self.input_files:
			input_parms+=str(file_name)+" "
		
		subprocess.call('plastimatch add ' + input_parms + str(self.output_file),\
		shell=True, stdout=self.log, stderr=self.log)
		
		self.log.close()




class adjust:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
	
	option={}
	
	_adjust_keys=("input", "output", "output-type", "scale", "ab-scale",\
	"stretch", "truncate-above", "truncate-below")
	
	def run_adjust(self):
		_run_plm_command("adjust", self.option, self._adjust_keys, self.log_file)




class convert:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
	
	option={}
	
	_convert_keys=("input","default-val","dif","dim","fixed",\
	"input-cxt","input-dose-ast","input-dose-img","input-dose-mc",\
	"input-dose-xio","input-ss-img","input-ss-list","interpolation",\
	"metadata","origin","output-color_map","output-ctx","output-dicom",\
	"output-dij","output-dose_img","output-img","output-labelmap",\
	"output-pointset","output-prefix","output-prefix_fcsv","output-ss_img",\
	"output-ss_list","output-type","output-vf","output-xio","patient-id",\
	"patient-name","patient-pos","prune-empty","referenced-ct","simplify-perc",\
	"spacing","xf","xor-contours")
	
	def run_convert(self):
		_run_plm_command("convert", self.option, self._convert_keys, self.log_file)




class crop:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
		
	option={}
	
	_crop_keys=("input", "output", "voxels")
	
	def run_crop(self):
		_run_plm_command("crop", self.option, self._crop_keys, self.log_file)




class diff:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
	
	input_files=[]
	output_file=""
	
	def run_diff(self):
		
		if len(self.input_files) != 2:
			raise NameError("You must define two input images!")
		
		if self.log_file == "":
			self.log=open(os.devnull, "w")
		else:
			self.log=open(self.log_file, "w")
		
		subprocess.call('plastimatch diff ' + str(self.input_files[0]) + " " + str(self.input_files[1])\
		+" " + str(self.output_file), shell=True, stdout=self.log, stderr=self.log)
		
		self.log.close()




class fill:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
	
	option={}
	
	_fill_keys=("input", "mask", "mask-value", "output", "output-format", "output-type")
	
	def run_fill(self):
		_run_plm_command("fill", self.option, self._fill_keys, self.log_file)




class mask:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
	
	option={}
	
	_mask_keys=("input", "mask", "mask-value", "output", "output-format", "output-type")
	
	def run_mask(self):
		_run_plm_command("mask", self.option, self._mask_keys, self.log_file)




class register:
	
	def __init__ (self, par_file="", log_file=""):
		self.par_file=par_file
		self.log_file=log_file
	
	_stage_keys=("xform","optim","impl","background_val","convergence_tol",\
	"demons_acceleration","demons_filter_width","demons_homogenization","demons_std",\
	"histoeq","grad_tol","grid_spac","max_its","max_step","metric","mi_histogram_bins",\
	"min_its","min_step","num_saples","regularization_lambda","res","ss","ss_fixed",\
	"ss_moving","threading","xform_in","xform_out","vf_out","img_out","img_out_fmt","img_out_type")
	
	_global_keys=("fixed","moving","xform_in","xform_out","vf_out","img_out",\
	"img_out_fmt","img_out_type","background_max")
	
	stages=[]
	_global_stage_added=False
	
	def add_global_stage(self):
		if self.par_file=="" or os.path.exists(self.par_file):
			raise NameError("GLOBAL STAGE NOT ADDED! You have to define a new parameters file name")
		else:
			if self._global_stage_added==False:
				self.stages=[{}]+self.stages
				self._global_stage_added=True
			else:
				raise NameError("The global stage already exists!")
	
	def add_stage(self):
		if self.par_file=="" or os.path.exists(self.par_file):
			raise NameError("STAGE NOT ADDED! You have to define a new parameters file name")
		else:
			self.stages+=[{}]
	
	def delete_stage(self, stage_number):
		if self.par_file=="" or os.path.exists(self.par_file):
			raise NameError("STAGE NOT DELETED! You have to define a new parameters file name")
		else:
			if stage_number != 0:
				del self.stages[stage_number]
			else:
				raise NameError("GLOBAL STAGE NOT DELETED! You can not delete the global stage")
	
	def run_registration(self):
		if not os.path.exists(self.par_file) and self.par_file!="":
			f=open(self.par_file, "w")			
			for stage_index, stage in enumerate(self.stages):
				if stage_index==0:
					stage=_clean_parms(stage, self._global_keys)
					f.write("[GLOBAL]\n")
				else:
					stage=_clean_parms(stage, self._stage_keys)
					f.write("\n[STAGE]\n")
				
				for key, value in dict.items(stage):
						f.write(key+"="+value+"\n")	
			f.close()
		
		if self.log_file == "":
			self.log=open(os.devnull, "w")
		else:
			self.log=open(self.log_file, "w")
		
		if self.par_file!="" and os.path.exists(self.par_file):
			print ("Please wait...")
			subprocess.call('plastimatch register ' + self.par_file, shell=True, stdout=self.log, stderr=self.log)
			self.log.close()
		else:
			raise NameError("REGISTRATION NOT EXECUTED! You have to define a new parameters file name")




class resample:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
		
	option={}
	
	_resample_keys=("default-value", "dim", "fixed", "input", "interpolation",\
	"origin", "output", "output-type", "spacing", "subsample")
	
	def run_resample(self):
		_run_plm_command("resample", self.option, self._resample_keys, self.log_file)




class segment:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
		
	option={}
	
	_segment_keys=("bottom", "debug", "fast", "input", "lower-treshold", "output-img")
	
	def run_segment(self):
		_run_plm_command("segment", self.option, self._segment_keys, self.log_file)




class warp:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
		
	option={}
	
	_warp_keys=("input","default-val","dif","dim","fixed",\
	"input-cxt","input-dose-ast","input-dose-img","input-dose-mc",\
	"input-dose-xio","input-ss-img","input-ss-list","interpolation",\
	"metadata","origin","output-color_map","output-ctx","output-dicom",\
	"output-dij","output-dose_img","output-img","output-labelmap",\
	"output-pointset","output-prefix","output-prefix_fcsv","output-ss_img",\
	"output-ss_list","output-type","output-vf","output-xio","patient-id",\
	"patient-name","patient-pos","prune-empty","referenced-ct","simplify-perc",\
	"spacing","xf","xor-contours")
	
	def run_warp(self):
		_run_plm_command("warp", self.option, self._warp_keys, self.log_file)




class xfconvert:
	
	def __init__ (self, log_file=""):
		self.log_file=log_file
	
	option={}
	
	_xfconvert_keys=("dim", "grid-spacing", "input", "nobulk", "origin",\
	"output", "output-type", "spacing")
	
	def run_xfconvert(self):
		_run_plm_command("xf-convert", self.option, self._xfconvert_keys, self.log_file)




####################### PUBLIC CLASSES - END - #########################




#################### UTILITY FUNCTION - START - ########################
############ PRIVATE UTILITY FUNCTION, NOT FOR PUBLIC USE ##############

def _clean_parms (d, t):
	
	return dict((k, v) for k, v in d.iteritems() if k in t)


def _run_plm_command(command_type, command_options, command_keys, command_log_file):
	
	if command_log_file == "":
		log=open(os.devnull, "w")
	else:
		log=open(command_log_file, "w")
	
	subprocess.call('plastimatch '+ command_type + _scan_options(command_options, command_keys),\
	shell=True, stdout=log, stderr=log)
	
	log.close()


def _scan_options (d, t):
	
		d=_clean_parms(d, t)
		
		special_keys=("voxels", "scale", "ab-scale", "stretch", "dim",\
		"grid-spacing", "origin", "spacing")
		
		opt_str=""
		
		for key, value in dict.items(d):
			if value!="Enabled" and value!="Disabled" and key not in special_keys:
				opt_str+=" --"+key+"="+value
			elif key in special_keys:
				opt_str+=" --"+key+"="+'"'+value+'"'
			elif value=="Enabled":
				opt_str+=" --"+key
			elif value == "Disabled":
				pass				
		
		return opt_str


############ PRIVATE UTILITY FUNCTION, NOT FOR PUBLIC USE ##############
##################### UTILITY FUNCTION - END - #########################
